### DISCUSSION TITLE
Discuss about better API for `nonlinsolve`

### DISCUSSION THREAD

**Discussion by MaqOwais** [sympy/sympy#21184] [Ideas]
I've not done a workout around this issue yet (but I'll try to follow up later on this ). I think it's very important and should be discussed about the output API of `nonlinsolve`.

**These are the reference links to start:** 

https://github.com/sympy/sympy/issues/11759#issuecomment-564563277

[discussion on mail about the inconsistency of api](https://groups.google.com/g/sympy/c/TMw5MmIO-qY/m/OXJ0oeVqAwAJ)

CREDITS: @oscarbenjamin

**Comment by asmeurer** [comment]
The problem is that the solution to a nonlinear system of equations may be complicated in general. It may depend on parameters (this is actually also true for linear systems, if they are under determined). Sets allow always representing solutions, even if they are complicated.

I think for solveset in general, it may be better to have some helper methods on the set object to convert common solution types into more useful data structures like dicts or lists. They wouldn't apply to generic set solutions, but would apply to a large class of solution sets.

**Reply by MaqOwais** [reply]
> I think for solveset in general, it may be better to have some helper methods on the set object to convert common solution types into more useful data structures like dicts or lists.

I think `solvify` is the function that calls the `solveset` function and returns the solution in terms of `list` same as `solve` API.
So I think in a similar fashion there should be a `dict` (as in `solve`) for multiple variables. For `nonlinsolve` 

**CASE - I**
```
>>> sys = [2*x + 2*y - 3, x + 3*y - 5]
>>> solve(sys,[x,y])
{x: -1/4, y: 7/4}
>>> nonlinsolve(sys, x,y)
FiniteSet((-1/4, 7/4))

# rather IMHO it could be
>>> nonlinsolve(sys, x,y)
{x: FiniteSet(-1/4), y:FiniteSet(7/4)}

```
**CASE - II**
```
>>> sys = [2*x + 2*y - 3, 2*sin(y) - 1]
>>> solve(sys)
[{x: 3/2 - 5*pi/6, y: 5*pi/6}, {x: 3/2 - pi/6, y: pi/6}]

>>> nonlinsolve(sys, x,y)
FiniteSet((ImageSet(Lambda(_n, -2*_n*pi - 5*pi/6 + 3/2), Integers), ImageSet(Lambda(_n, 2*_n*pi + 5*pi/6), Integers)), 
(ImageSet(Lambda(_n, -2*_n*pi - pi/6 + 3/2), Integers), ImageSet(Lambda(_n, 2*_n*pi + pi/6), Integers)))

# rather IMHO it could be
>>> nonlinsolve(sys, x,y)
[ 
{x : ImageSet(Lambda(_n, -2*_n*pi - 5*pi/6 + 3/2), Integers),  y : ImageSet(Lambda(_n, 2*_n*pi + 5*pi/6), Integers)}, 
{x : ImageSet(Lambda(_n, -2*_n*pi - pi/6 + 3/2), Integers), y : ImageSet(Lambda(_n, 2*_n*pi + pi/6), Integers)}
]

```
